### PR TITLE
perf: only load possible results in UserMountCache::getMountForPath

### PR DESCRIPTION
**Pull request by icewind1991**
Extracted from https://github.com/nextcloud/server/pull/57760 for easier reviews.

Instead of always loading all mounts for the user and then finding the relevant one for a path. We first collect the possible paths the mountpoint can be at (the path and all it's parents), and then only load the mounts for those paths.

### CODE REVIEW

**Review by salmart-dev — APPROVED** [review]

**Review by CarlSchwan — APPROVED** [review]

File: lib/private/Files/Config/UserMountCache.php [hunk 1]
```diff
@@ -482,28 +482,48 @@ public function clear(): void {
...
-		foreach ($this->getMountsForUser($user) as $mount) {
-			$mounts[$mount->getMountPoint()] = $mount;
-		}
-
+		$searchPaths = [];
 		$current = rtrim($path, '/');
-		// walk up the directory tree until we find a path that has a mountpoint set
-		// the loop will return if a mountpoint is found or break if none are found
-		while (true) {
+		// get all paths that we are interested in, $path and all it's parents
```
Comment (joshtrichards) [inline lib/private/Files/Config/UserMountCache.php:504]: Super minor comment tweak:
```suggestion
		// get all paths that we are interested in, $path and all its parents
```

File: lib/private/Files/Config/UserMountCache.php [hunk 2]
```diff
@@ -482,28 +482,48 @@ public function clear(): void {
...
+				->where($builder->expr()->eq('user_id', $builder->createNamedParameter($user->getUID())))
+				->andWhere($builder->expr()->in('mount_point_hash', $builder->createNamedParameter($searchPathHashes, IQueryBuilder::PARAM_STR_ARRAY)));
+
+			foreach ($query->executeQuery()->fetchAll() as $row) {
+				$mount = $this->dbRowToMountInfo($row);
+				$mounts[$mount->getMountPoint()] = $mount;
+			}
+		}
+
+		// note that $searchPaths is sorted deepest path first
```
Comment (joshtrichards) [inline lib/private/Files/Config/UserMountCache.php:537]: Minor clarity:

```suggestion
		// $searchPaths is ordered deepest-first, so the first hit is the best match.
```

File: lib/private/Files/Config/UserMountCache.php [hunk 3]
```diff
@@ -482,28 +482,48 @@ public function clear(): void {
...
+		foreach ($searchPaths as $searchPath) {
+			if (isset($mounts[$searchPath])) {
```
Comment (joshtrichards) [inline lib/private/Files/Config/UserMountCache.php:539]: Readability: Maybe use `$candidateMounts` throughout instead of `$mounts` to make this final matching loop easier to parse iclearer these are the possible matches, not necessarily all user mounts.

```suggestion
			if (isset($candidateMounts[$searchPath])) {
```

File: lib/private/Files/Config/UserMountCache.php [hunk 4]
```diff
@@ -482,28 +482,48 @@ public function clear(): void {
...
 			}
 		}
 
+		$mounts = [];
+		if (isset($this->mountsForUsers[$user->getUID()])) {
+			foreach ($this->mountsForUsers[$user->getUID()] as $mount) {
+				$mounts[$mount->getMountPoint()] = $mount;
+			}
+		} else {
+			$searchPathHashes = array_map(static fn (string $path) => hash('xxh128', $path), $searchPaths);
```
Comment (joshtrichards) [inline lib/private/Files/Config/UserMountCache.php:522]: Readability: Maybe rename the closure arg from `$path` to `$candidatePath` to avoid reusing the overarching function parameter's existing parameter name.

File: tests/lib/Files/Config/UserMountCacheTest.php [hunk 5]
```diff
@@ -619,4 +625,48 @@ public function testChangedSameRootId(): void {
 		$mountIds = array_map(fn (ICachedMountInfo $mountInfo) => $mountInfo->getMountId(), $cached);
 		$this->assertEquals([null, 1], $mountIds);
 	}
+
+	public function testGetMountForPath(): void {
```
Comment (joshtrichards) [inline tests/lib/Files/Config/UserMountCacheTest.php:634]: Perhaps also a test for the negative / `NotFoundException` path?

**Review by joshtrichards — COMMENTED** [review]
Seems like a reasonable optimization. Just some minor comment/suggestions.
